### PR TITLE
Pass DeleteOptions down to the Reactor

### LINKED ISSUE #102798: Fake client should support mimic graceful pod deletion
#### What would you like to be added:
Fake client should support mimic graceful pod deletion. Although fake client supports `AddReactor/PrependReactor` to add  a customize `Reactor` to different requests, it doesn't pass DeleteOptions down to the `Reactor`.

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/kubernetes/typed/core/v1/fake/fake_pod.go#L121-L126

#### Why is this needed:
To mimic testing graceful pod deletion in my project.

### PR DESCRIPTION
**Pull request by chenchun**
#### What type of PR is this?

/kind feature

#### Which issue(s) this PR fixes:

Fixes #102798

#### Does this PR introduce a user-facing change?

```release-note
client-go: pass `DeleteOptions` down to the fake client `Reactor`
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/fake/generator_fake_for_type.go [hunk 1]
```diff
@@ -412,8 +412,8 @@ var deleteTemplate = `
 // Delete takes name of the $.type|private$ and deletes it. Returns an error if one occurs.
 func (c *Fake$.type|publicPlural$) Delete(ctx context.Context, name string, opts $.DeleteOptions|raw$) error {
 	_, err := c.Fake.
-		$if .namespaced$Invokes($.NewDeleteAction|raw$($.type|allLowercasePlural$Resource, c.ns, name), &$.type|raw${})
-		$else$Invokes($.NewRootDeleteAction|raw$($.type|allLowercasePlural$Resource, name), &$.type|raw${})$end$
+		$if .namespaced$Invokes($.NewDeleteAction|raw$($.type|allLowercasePlural$Resource, c.ns, name, &opts), &$.type|raw${})
+		$else$Invokes($.NewRootDeleteAction|raw$($.type|allLowercasePlural$Resource, name, &opts), &$.type|raw${})$end$
```
Comment (enj) [inline staging/src/k8s.io/code-generator/cmd/client-gen/generators/fake/generator_fake_for_type.go:416]: This needs to pass in a `nil` when `opts` is the zero value of the struct.

File: staging/src/k8s.io/client-go/testing/actions.go [hunk 2]
```diff
@@ -221,11 +221,12 @@ func NewUpdateSubresourceAction(resource schema.GroupVersionResource, subresourc
 	return action
 }
 
-func NewRootDeleteAction(resource schema.GroupVersionResource, name string) DeleteActionImpl {
+func NewRootDeleteAction(resource schema.GroupVersionResource, name string, opts *metav1.DeleteOptions) DeleteActionImpl {
```
Comment (enj) [inline staging/src/k8s.io/client-go/testing/actions.go:224]: Update `DeleteActionImpl` and `DeleteAction` to make it easier to get the options:

```go
type DeleteAction interface {
	Action
	GetName() string
	GetDeleteOptions() *metav1.DeleteOptions
}
```

**Review by enj — CHANGES_REQUESTED** [review]

**Review by enj — CHANGES_REQUESTED** [review]
Minor stuff.

**Review by enj — CHANGES_REQUESTED** [review]
Just some extraneous whitespace.

File: pkg/controller/bootstrap/common_test.go [hunk 3]
```diff
@@ -34,6 +34,7 @@ func newTokenSecret(tokenID, tokenSecret string) *v1.Secret {
 			Namespace:       metav1.NamespaceSystem,
 			Name:            bootstrapapi.BootstrapTokenSecretPrefix + tokenID,
 			ResourceVersion: "1",
+			UID:             "s1",
```
Comment (liggitt) [inline pkg/controller/bootstrap/common_test.go:37]: make the uid unique by incorporating tokenID (`"uid"+tokenID` or something similar)

**Review by liggitt — COMMENTED** [review]
one nit on the uid used in the test, lgtm otherwise

**Pull request merged**